core/templates/pages/exploration-player-page/services/command-executor.service.ts

```diff
@@ -0,0 +1,210 @@
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Oppia interaction from outer iframe with messaging commands.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import {WindowRef} from 'services/contextual/window-ref.service.ts';
+import {WindowWrapperMessageService} from
+  'pages/exploration-player-page/services/window-wrapper-message.service';
+require('pages/exploration-player-page/services/exploration-engine.service.ts');
+import {ServicesConstants} from 'services/services.constants';
+
+@Injectable({providedIn: 'root'})
+export class CommandExecutorService {
+  private commandToFunctionMap = {
+    CONTINUE: this.continueClick,
+    ENTER_TEXT_NUMBER_UNITS: this.fillTextBox,
+    ADD_SET: this.addSet,
+    REMOVE_SET: this.removeSet,
+    ENTER_FRACTION: this.enterFraction,
+    SELECT_ITEM_BULLET: this.selectItemBullet,
+    SUBMIT: this.submit,
+  };
+  setElementsOnPage = 0;
+  hostname = '';
+  cachedOuterFrameMessage = '';
+  windowWrapperMessageService = null;
+  constructor(private windowRef: WindowRef) {
+    this.setElementsOnPage = 0;
+    this.hostname = '';
+    this.cachedOuterFrameMessage = '';
+    this.windowWrapperMessageService =
+      new WindowWrapperMessageService(windowRef);
+  }
+
+  sendParentReadyState(windowRef) {
+    this.windowWrapperMessageService.postMessageToParent(
+      'Ready to receive hostname', '*');
+  }
+
+  getOuterFrameEvents(windowRef) {
+    this.windowWrapperMessageService.addEventListener(
+      'message', (event) => {
+        var messageArray = event.data.split(' ');
+        var command = messageArray[0];
+        var message = '';
+        for (var i = 1; i < messageArray.length; i++) {
+          message = message + messageArray[i] + ' ';
+        }
+        message = message.substr(0, message.length - 1);
+        if ((command !== 'HOSTNAME' && this.hostname === '') ||
+        command === 'finishedProcessing') {
+          return;
+        } else if (command === 'CONTINUE' || command === 'SUBMIT') {
+          this.commandToFunctionMap[command](windowRef);
+        } else if (command === 'HOSTNAME') {
+          if (ServicesConstants.WHITELISTED_IFRAME_HOSTS.indexOf(message) >=
+          0) {
+            this.hostname = message;
+          }
+          if (this.cachedOuterFrameMessage !== '' &&
+          this.cachedOuterFrameMessage !== undefined) {
```

Comment:
Should we be resetting the outgoing event after sending it? Also, why are we sending it after receiving other commands—why not immediately? What’s the criteria for deciding when to send it?

Reply:
So we first have a handshake protocol-- in which oppia tells all parents that it is ready to begin receiving messages. Then the messager replies with its hostname. Until this point, oppia cannot send interaction state information to parents because the hostname is not known. In situations where the interaction state is changed before the hostname is sent, the state is cached. Then when the hostname is received, oppia immediately sends the hostname the cached interaction. This is the criteria.